pandas/core/dtypes/cast.py

```diff
@@ -1505,12 +1505,20 @@ def cast_scalar_to_array(shape, value, dtype: Optional[DtypeObj] = None) -> np.n
 
     """
     if dtype is None:
-        dtype, fill_value = infer_dtype_from_scalar(value)
+        dtype, fill_value = infer_dtype_from_scalar(value, pandas_dtype=True)
     else:
         fill_value = value
 
-    values = np.empty(shape, dtype=dtype)
-    values.fill(fill_value)
+    # TODO: Update this function to add support for 3rd party extension types
+    # Issue #34959
+    if is_extension_array_dtype(dtype):
+        if isinstance(shape, int):
+            shape = (shape, 1)
+        value = [construct_1d_arraylike_from_scalar(value, shape[0], dtype)]
+        values = value * shape[1]
```

Comment:
@justinessert this results in a dataframe where all the columns are the same object. So that something like

```python
df = pd.DataFrame(pd.Timetsamp('2000', tz='CET'), index=[0, 1], columns=[0, 1])
df.iloc[0, 0] = pd.Timestamp("2001")
```

would mutate every column. Try

```python
value = [construct_1d_arraylike_from_scalar(value, shape[0], dtype) for _ in range(shape)]
```

Reply:
Sorry @TomAugspurger I don't understand, this works as I would expect it to. Can you rephrase the issue?

My testing:

```python
df = pd.DataFrame(pd.Timestamp('2000', tz='CET'), index=[0, 1], columns=[0, 1])
```

^Works as expected, both columns 0 and 1 are `datetime64[ns, CET]`

```python
df1 = df.copy()
df1.iloc[0,0] = pd.Timestamp("2001")
```

^This one shouldn't even call the cast scalar to array function I believe, since it is only assigning a single value, not a full column. But this also works as expected for me, column 1 remains `datetime64[ns, CET]` and column 0 becomes `object` (since it has mixed types).

Other similar cases:

```python
df2 = df.copy()
df2[0] = pd.Timestamp("2001")
```

^This does call cast_scalar_to_array and also works as expected I believe. Column 0 becomes `datetime64[ns]` and column 1 remains `datetime64[ns, CET]`

```python
df3 = df.copy()
df3.iloc[0] = pd.Timestamp("2001")
```

^This does call cast_scalar_to_array and also works as expected I believe. Both columns become `object` types (because they have mixed types)

Reply:
Sorry, bad example. The value being set should have had the same timezone. Does this help?

```python
In [13]: df = pd.DataFrame(pd.Timestamp('2000', tz='CET'), index=[0, 1], columns=[0, 1])
    ...:

In [14]: df[0].array is df[1].array
Out[14]: True

In [16]: df[0].array[0] = pd.Timestamp("2001", tz="CET")

In [17]: df
Out[17]:
                          0                         1
0 2001-01-01 00:00:00+01:00 2001-01-01 00:00:00+01:00
1 2000-01-01 00:00:00+01:00 2000-01-01 00:00:00+01:00
```

Notice that the `1` column was mutated too. We want to create a distinct array for each column, not a single array each column refers to.

Reply:
Ahh, yes that makes complete sense. Updated